### PR TITLE
test: [Photo] UseCase 단위 테스트 추가

### LINKED ISSUE #252: test: [Photo] 도메인 UseCase 단위 테스트 (~52 tests)
## UseCase Unit Test - Photo 도메인 (~62 tests)

가장 많은 UseCase (14개). cascade 삭제, 업로드 롤백 등 복잡한 트랜잭션 포함.

### 선행 조건
- #249 (테스트 인프라) 머지 필요

---

### 1. CreateFolderUseCase

> **소스**: `src/main/kotlin/com/neki/photo/application/usecase/CreateFolderUseCase.kt`
> **의존성**: `FolderRepositoryPort` (1개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 생성 | 이름 고유 → save + ID 반환 |
| 2 | 이름 충돌 | `BusinessException(CONFLICT_FOLDER)` |

---

### 2. UpdateFolderUseCase

> **의존성**: `FolderRepositoryPort` (1개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 수정 | 폴더 존재 + 새 이름 고유 → update |
| 2 | 폴더 미존재 | `BusinessException(NOT_FOUND)` |
| 3 | 현재 이름과 동일 | 충돌 검사 skip, 정상 처리 |
| 4 | 새 이름 충돌 | `BusinessException(CONFLICT_FOLDER)` |

---

### 3. PutPhotoUseCase

> **의존성**: `PhotoImageRepositoryPort` (1개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 수정 | photo 존재 → update 호출 |
| 2 | 미존재 | `BusinessException(NOT_FOUND)` |
| 3 | **memo가 null** | null memo → 기존 값 유지 vs 클리어 동작 확인 |

---

### 4. UpdatePhotoFavoriteUseCase

> **의존성**: `PhotoImageRepositoryPort`, `FavoriteImageRepositoryPort` (2개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 즐겨찾기 추가 (favorite=true) | `add()` 호출 |
| 2 | 즐겨찾기 해제 (favorite=false) | `delete()` 호출 |
| 3 | 사진 미존재 | `BusinessException` |
| 4 | **이미 즐겨찾기인데 add** | 멱등성 — repository 동작 확인 |
| 5 | **즐겨찾기 아닌데 delete** | 멱등성 — repository 동작 확인 |

---

### 5. UpdatePhotoUseCase (deprecated)

> **의존성**: `PhotoImageRepositoryPort` (1개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 수정 | update 호출 확인 |
| 2 | 미존재 | `BusinessException(NOT_FOUND)` |
| 3 | **memo가 null** | `command.memo?.let` — null이면 기존 값 유지 확인 |

---

### 6. GetFoldersUseCase

> **의존성**: `FolderRepositoryPort` (1개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 조회 | 폴더 목록 반환 |
| 2 | 빈 목록 | 빈 리스트 |
| 3 | **coverImageStorageKey가 null** | 커버 이미지 없는 폴더 → null 반환 확인 |

---

### 7. GetPhotosUseCase

> **의존성**: `PhotoImageRepositoryPort`, `MediaClientPort` (2개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 조회 + 미디어 매핑 | photos + storageInfo 매핑 |
| 2 | hasNext=true | photos.size > command.size |
| 3 | hasNext=false | 마지막 페이지 |
| 4 | 빈 결과 | 빈 리스트 + hasNext=false |
| 5 | 미디어 미존재 (eventual consistency) | 해당 photo 필터링 |
| 6 | **모든 사진의 미디어 미존재** | hasNext=true이지만 반환 리스트 비어있는 케이스 |

---

### 8. GetFavoritePhotosUseCase

> **의존성**: `PhotoImageRepositoryPort`, `MediaClientPort` (2개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 조회 | 즐겨찾기 photos + 미디어 매핑 |
| 2 | hasNext=true | 다음 페이지 존재 |
| 3 | hasNext=false | 마지막 페이지 |
| 4 | 빈 결과 | 즐겨찾기 없음 |
| 5 | **모든 미디어 미존재** | 데이터 있지만 미디어 아직 없음 → 빈 결과 반환 |

---

### 9. GetPhotoUseCase

> **의존성**: `PhotoImageRepositoryPort`, `MediaClientPort` (2개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 조회 | photo + media storageInfo 매핑 |
| 2 | 미존재 | `BusinessException(NOT_FOUND)` |
| 3 | **사진 존재하지만 미디어 조회 실패** | mediaClient 예외 → 전파 확인 |

---

### 10. GetFavoriteSummaryUseCase

> **의존성**: `FavoriteImageRepositoryPort`, `PhotoImageRepositoryPort`, `MediaClientPort` (3개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | count=0 | storageKey=null, count=0 |
| 2 | count>0, 최신 photo 없음 (edge) | storageKey=null |
| 3 | count>0, photo 있으나 media 미존재 | storageKey=null |
| 4 | 정상 | count + storageKey 반환 |
| 5 | **count>0이지만 latestPhoto null** | 즐겨찾기 수와 실제 photo 간 불일치 (race condition) → log.warn + null storageKey |

---

### 11. RemovePhotosFromFolderUseCase

> **의존성**: `FolderRepositoryPort`, `PhotoImageRepositoryPort`, `PhotoImageFolderRepositoryPort` (3개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 삭제 | 폴더 소유 확인 → dual-write 삭제 |
| 2 | 소유권 없음 | `BusinessException` |
| 3 | **빈 photoIds 리스트** | 빈 목록으로 호출 → 멱등 동작 확인 |

---

### 12. DeletePhotosUseCase

> **의존성**: `PhotoImageRepositoryPort`, `FavoriteImageRepositoryPort`, `MediaClientPort`, `TransactionRunner` (4개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 삭제 | 즐겨찾기 삭제 → 사진 삭제 → media cleanup 호출 |
| 2 | 삭제할 사진이 빈 리스트 | media cleanup 미호출 |
| 3 | 미존재 사진 | `BusinessException` |
| 4 | **media cleanup 실패** | DB 삭제 성공 → mediaClient 예외 → 미디어 고아 상태 (트랜잭션 밖) |

---

### 13. DeleteFoldersUseCase

> **의존성**: `FolderRepositoryPort`, `PhotoImageRepositoryPort`, `PhotoImageFolderRepositoryPort`, `FavoriteImageRepositoryPort`, `MediaClientPort`, `TransactionRunner` (6개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | deletePhotos=true | 즐겨찾기 → 사진 → 폴더 삭제 + media cleanup |
| 2 | deletePhotos=false | folderId null 처리 + 폴더 삭제, media 미호출 |
| 3 | 폴더 카운트 불일치 | `BusinessException(NOT_FOUND)` |
| 4 | 빈 폴더 + deletePhotos=true | photoIdsToDelete 비어있음 → media cleanup 미호출 |
| 5 | **media cleanup 실패** | DB 삭제 성공 → mediaClient 예외 → 미디어 고아 상태 (트랜잭션 밖) |
| 6 | **deletePhotos=false 시 사진 folderId null 처리** | updatePhotosFolderIdToNull 호출 확인 |

---

### 14. UploadPhotosUseCase

> **의존성**: `MediaClientPort`, `PhotoImageRepositoryPort`, `PhotoImageFolderRepositoryPort`, `FolderRepositoryPort`, `FavoriteImageRepositoryPort`, `TransactionRunner` (6개)

| # | 케이스 | 검증 |
|---|--------|------|
| 1 | 정상 업로드 | 미디어 확인 → 사진 저장 + 폴더 연결 + 즐겨찾기 |
| 2 | 중복 mediaId | `BusinessException` |
| 3 | 폴더 미소유 | `BusinessException` |
| 4 | 일부 미디어 미사용 가능 | 롤백 처리 |
| 5 | 트랜잭션 실패 → 롤백 | 미디어 롤백 호출 확인 |
| 6 | 멱등성 - 기존 mediaId 필터링 | 중복 제거 후 저장 |
| 7 | 모든 mediaId 이미 존재 | early return |
| 8 | **folderId가 null** | 폴더 없이 사진 업로드 → photoImageFolder 생성 skip |
| 9 | **ALREADY_REQUEST 예외** | saveAll에서 ALREADY_REQUEST → 롤백 없이 early return |
| 10 | **롤백 중 예외 발생** | 트랜잭션 실패 → rollback도 예외 → 원래 예외 마스킹 여부 |

---

## 완료 조건

- [ ] CreateFolderUseCase 테스트
- [ ] UpdateFolderUseCase 테스트
- [ ] PutPhotoUseCase 테스트
- [ ] UpdatePhotoFavoriteUseCase 테스트
- [ ] UpdatePhotoUseCase 테스트
- [ ] GetFoldersUseCase 테스트
- [ ] GetPhotosUseCase 테스트
- [ ] GetFavoritePhotosUseCase 테스트
- [ ] GetPhotoUseCase 테스트
- [ ] GetFavoriteSummaryUseCase 테스트
- [ ] RemovePhotosFromFolderUseCase 테스트
- [ ] DeletePhotosUseCase 테스트
- [ ] DeleteFoldersUseCase 테스트
- [ ] UploadPhotosUseCase 테스트
- [ ] `./gradlew test` 통과

### PR DESCRIPTION
**Pull request by koosco**
## Summary
14개 UseCase 61개 테스트 추가:
- CreateFolderUseCase(2), UpdateFolderUseCase(4), PutPhotoUseCase(3), UpdatePhotoFavoriteUseCase(5)
- UpdatePhotoUseCase(3), GetFoldersUseCase(3), GetPhotosUseCase(6), GetFavoritePhotosUseCase(5)
- GetPhotoUseCase(3), GetFavoriteSummaryUseCase(5), RemovePhotosFromFolderUseCase(3)
- DeletePhotosUseCase(3), DeleteFoldersUseCase(6), UploadPhotosUseCase(10)

cascade 삭제, 업로드 롤백, 미디어 고아 상태, BaseTimeEntity NPE 등 edge case 포함

closes #252

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/test/kotlin/com/neki/photo/application/usecase/UploadPhotosUseCaseTest.kt [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        every { photoImageRepository.saveAll(any()) } throws originalException
+        every { mediaClient.rollbackMediasUploaded(1L, listOf(10L)) } throws RuntimeException("롤백 오류")
+
+        // When & Then - 롤백 중 예외가 발생해도 원래 예외가 전파됨
+        val ex = shouldThrow<RuntimeException> {
+            useCase.execute(command)
+        }
+        // 원래 예외가 전파되어야 함 (롤백 예외에 마스킹될 수 있음)
+        verify(exactly = 1) { mediaClient.rollbackMediasUploaded(1L, listOf(10L)) }
+    }
```
[the remark this answers is not in the document]
Comment (koosco) [reply src/test/kotlin/com/neki/photo/application/usecase/UploadPhotosUseCaseTest.kt]: 반영 완료했습니다. 롤백 중 예외 발생 시 실제로 전파되는 예외가 rollbackException임을 확인하고, `ex shouldBe rollbackException` 으로 구체적인 예외 동일성 검증을 추가했습니다. (6c9ea0b)

File: src/test/kotlin/com/neki/photo/application/usecase/GetFavoriteSummaryUseCaseTest.kt [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        every { favoriteImageRepository.countByUserId(1L) } returns 2L
+        every { photoImageRepository.getLatestFavoritePhoto(1L) } returns null
+
+        // When
+        val result = useCase.execute(command)
+
+        // Then
+        result.storageKey shouldBe null
+        result.totalCount shouldBe 2L
+    }
```
[the remark this answers is not in the document]
Comment (koosco) [reply src/test/kotlin/com/neki/photo/application/usecase/GetFavoriteSummaryUseCaseTest.kt]: 반영 완료했습니다. 109번 라인의 "count>0이지만 latestPhoto가 null인 경우" 테스트가 48번 라인과 동일한 시나리오를 검증하므로 중복 케이스를 제거했습니다. (6c9ea0b)

File: src/test/kotlin/com/neki/photo/application/usecase/UpdatePhotoFavoriteUseCaseTest.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        every { photoImageRepository.existsOwnedPhoto(1L, 1L) } returns true
+        every { favoriteImageRepository.delete(1L, 1L) } just Runs
+
+        // When
+        useCase.execute(command)
+
+        // Then
+        verify(exactly = 1) { favoriteImageRepository.delete(1L, 1L) }
+    }
```
[the remark this answers is not in the document]
Comment (koosco) [reply src/test/kotlin/com/neki/photo/application/usecase/UpdatePhotoFavoriteUseCaseTest.kt]: 반영 완료했습니다. UpdatePhotoFavoriteUseCase는 상태를 조회하지 않고 단순히 repository 명령을 호출하므로 74번, 88번 라인의 "이미 즐겨찾기인 상태" 및 "즐겨찾기가 아닌 상태" 케이스가 28번, 44번 라인과 동일한 로직을 검증하는 중복 테스트였습니다. 두 케이스를 제거했습니다. (6c9ea0b)

File: src/test/kotlin/com/neki/photo/application/usecase/DeleteFoldersUseCaseTest.kt [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+        every { photoImageFolderRepository.deleteByFolderIds(folderIds) } just Runs
+        every { folderRepository.deleteOwnedFolders(1L, folderIds) } returns 2
+
+        // When
+        useCase.execute(command)
+
+        // Then
+        verify(exactly = 1) { photoImageRepository.updatePhotosFolderIdToNull(1L, folderIds) }
+        verify(exactly = 0) { photoImageRepository.deleteOwnedPhotos(any(), any()) }
+    }
```
[the remark this answers is not in the document]
Comment (koosco) [reply src/test/kotlin/com/neki/photo/application/usecase/DeleteFoldersUseCaseTest.kt]: 반영 완료했습니다. 164번 라인의 "deletePhotos=false이면 updatePhotosFolderIdToNull 호출 확인" 테스트가 73번 라인의 테스트와 동일한 시나리오를 검증하므로 중복 케이스를 제거했습니다. (6c9ea0b)
